### ISSUE TITLE
Sched support for RFC 14

### ISSUE THREAD

**Issue by lipari** [flux-framework/flux-sched#192]
The following is a rough plan for enhancing the flux-sched scheduler to handle job requests that conform to the Canonical Job Specification [RFC 14](https://github.com/flux-framework/rfc/blob/master/spec_14.adoc).  Please provide feedback to clarify, correct, or extend the planned sequence of steps outlined below.

The existing sequence for processing a job spec goes like this:

* Job requirements are passed as arguments (node count, task count, time, etc.) to `flux-submit`
* `flux-submit` creates a lwj entry in the kvs that captures those arguments
* The sched module receives an event announcing the new job
* sched invokes the Job Status and Control (JSC) API to retrieve specific fields about that job
* sched builds a `flux_lwj_t` object with data from those fields saved to its `flux_res_t` member
* sched queues the `flux_lwj_t` job object
* sched's schedule_jobs() cycles through the queued job objects, decomposes the `flux_lwj_t/flux_res_t` object and creates a json object with predefined fields (e.g. "nodes", "starttime", etc.)
* sched then invokes `resrc_reqst_from_json()` to create a `resrc_reqst_t` (resource request) object
* This `resrc_reqst_t` object is passed to sched's `find()` and `select()` API calls to create the allocate-able (or reserve-able) resources.

While this mechanism has served as an initial prototype to get things going, it needs to be modified to conform to [RFC 14](https://github.com/flux-framework/rfc/blob/master/spec_14.adoc).  The proposed solution would be to reduce the above steps to the following:

* Job requirements are passed as arguments to `flux-submit`
* `flux-submit` creates a lwj entry in the kvs which captures those arguments in the canonical job spec format. (on-going work by @trws )
* sched invokes the Job Status and Control (JSC) API to retrieve the canonical job spec
* sched translates the canonical job spec to a resrc_reqst_t object
* (As above) this resrc_reqst_t object is passed to sched's `find()` and `select()` API calls to create the allocate-able (or reserve-able) resources.

**Comment by garlick** [comment]
Could the scheduler's find and select calls eventually accept the canonical jobspec?  Just thinking we might have a similar situation to the one discussed in flux-framework/flux-core#801 WRT KVS treeobj objects.  If we've defined the jobspec in an RFC, should the end-goal be to use it ubiquitously?

**Comment by dongahn** [comment]
@lipari: thank you for pointing all the changes we will ultimately have to make. 

Some questions: Canonical job specs will be primarily used both by wreck (and its replacement) and sched? Maybe, an alternative process is to start an effort to separate resrc and its friend from `sched` and make it the public canonical job and resource API? This ways both can be married a bit more tightly. 

Then, Incremental transitions of `sched` you describe can also help us firm up the shiny new canonical job and resource API?

**Comment by lipari** [comment]
> Could the scheduler's find and select calls eventually accept the canonical jobspec?

Yes.  The current resrc search implementation works best using a resrc_reqst_t object.  So, for the short term, a canonical spec -> resrc_reqst_t translation is expedient.

If we replace the current resrc implementation with something different, then I could see the request object becoming the canonical job spec.

> Maybe, an alternative process is to start an effort to separate resrc and its friend from sched and make it the public canonical job and resource API?

Yes, we bundled the resource representation into the sched project for convenience.  Now, I can see merits to breaking RDL/resrc (and flux-submit) out into its own stand-alone project that serves flux-core and flux-sched - or replacing RDL/resrc entirely with a new project.

**Comment by dongahn** [comment]
@lipari: 

A related, but an off-topic. 

As part of the new API, I think it would be helpful if we can expose a "transaction" pattern as the first class abstraction as well. 

I think that a key to scheduler scalability would be an ability to prune unnecessary tree traversal. So, if this abstraction can allow a scheduler to easily "speculate" a resource allocation, update its aggregates  (e.g., available cores available underneath me), and either commit or rollback efficiently at the end, this can be a big win.

**Comment by SteVwonder** [comment]
@dongahn: agreed!  My experience with IO-Aware scheduling aligns with suggestion.

A big performance win for IO-Aware scheduling is checking how much IO BW is available at the PFS and comparing that to (# nodes * IO BW per node).  It saves lots of tree traversing by allowing a quick bailout if the job requests more aggregate IO than the PFS can provide.

**Comment by lipari** [comment]
At long last, I am finally getting back to thinking about converting flux-sched to use the canonical job spec.  What follows is a rough outline of the tasks as I see them...

Here is the historical discussion regarding the canonical job/resource spec:

- https://github.com/flux-framework/flux-core/issues/354

- https://github.com/flux-framework/rfc/pull/29

- https://github.com/flux-framework/rfc/issues/55

And here is the commit that added the canonical job spec to RFC 14:

- https://github.com/flux-framework/rfc/pull/51

Here is how I breakdown the remaining work:

- Implement the canonical job spec as defined in RFC 14.  This would be, in effect, a new resrc.  I'd like to make this its own flux-framework project:  flux-resource (we'll throw in some vowels this time)

A large part of this effort will be devoted to evaluating and selecting the most appropriate solution for representing graph relationships.  This starts with the simple tree-like hardware (composite) hierarchy, aka "has-a" relationship.  But then, it grows to include aggregation relationships where one resource is connected in some way to another (e.g., power flow, network connectivity).

- Update and extend our existing tools for loading a set of resources into a flux-resource resource set:
	* From file (revising or replacing the RDL)
	* From hwloc
	* From a db?
	* From the KVS

- Develop a means for storing the complete flux-resource resource set:
	* Into the KVS
	* To a file
	* To a db?

- Create a set of tools for interrogating the resources stored in the KVS and add/delete/modifing them.  This would be the equivalent of "scontrol show node" and "sinfo".  Also, a command to display all the jobs running on nodes 57 through 89.

Next, I submit that there is a large overlap between representing resources and a job's request for resources.  The current code recognizes this commonality:  the resrc_reqst_t struct includes a resrc_t member.  So, next to do:

- Devise shorthand notation for representing a request for a set of canonical resources.  This is where we map flux-submit options/arguments to specific resource requests (e.g. "flux submit -N <xxx>" <==> "give me xxx nodes")

- Implement the job request for resources into a job object which gets stored in the KVS.  This greatly expands what we do now to include a representation of the canonical resource request.

Note that there is a subtle distinction between the job's request and what it is allocated.  The complete job object will need to provide this distinction.  Hence, there might be a field for a request of 4 - 16 nodes, in multiples of 4, and also a field for what was ultimately allocated to the job:  12 nodes along with the specific list of allocated nodes.

- Revise the existing "find", "select", "allocate" and "reserve" code to support the search, selection, and allocation of canonical resources.  This will impact the "planner" code @dongahn is working on.

On the flux-core side, there could also be major changes.  On one hand, I can see the elegance of keeping flux-core rank-aware only, with no knowledge of the resources each rank is running on.  Under this scenario, we need to maintain a "Job Status and Control" component to translate between ranks and the set of resources each rank runs on.

However, another option to consider is to bring resource-awareness into flux-core.  wreckrun's replacement, flux-run would become aware of more than just ranks.  It would leverage the resource set provided by the flux-resource project.  E.g., flux-run would need to support a request to "run 24 tasks across nodes:  47, 48, 63, and 75, with 8 tasks per node, distributed 4 tasks to a socket.

- Whatever we decide, we may need to revisit the Job Status and Control code.  In theory, getting job-to-resource mappings should be a simple call to retrieve job info from the KVS.

**Comment by dongahn** [comment]
@lipari: This sounds all good. Lifting the resrc and friend and move them into a separate flux-resource project as a standalone tool makes a lot of sense to me. 

Only high-level comment I have at the point, as you and I also discussed this before: It seem we need to coordinate this work closely with my internal scalability work + match/visitor patterning work through some good planning. As I can imagine, this can increase development cost if we don't untangle complexity in work depedencies involved in that surgery.

Welcome back! :-)
